x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -941,72 +1005,49 @@ public void onFailure(Exception e) {
+     **/
+    void decryptAndReturnSupersedingTokens(String refreshToken, RefreshTokenStatus refreshTokenStatus,
+                                           ActionListener<Tuple<String, String>> listener) {
+        final byte[] iv = Base64.getDecoder().decode(refreshTokenStatus.getIv());
+        final byte[] salt = Base64.getDecoder().decode(refreshTokenStatus.getSalt());
+        final byte[] encryptedSupersedingAccessToken = Base64.getDecoder().decode(refreshTokenStatus.getSupersedingAccessToken());
+        final byte[] encryptedSupersedingRefreshToken = Base64.getDecoder().decode(refreshTokenStatus.getSupersedingRefreshToken());
+        try {
+            Cipher cipher = getDecryptionCipher(iv, refreshToken, salt);
+            final String supersedingAccessToken = new String(cipher.doFinal(encryptedSupersedingAccessToken), StandardCharsets.UTF_8);
+            byte[] iv2 = new byte[iv.length];
+            System.arraycopy(iv, iv.length / 2, iv2, 0, iv.length / 2);
+            System.arraycopy(iv, 0, iv2, iv.length / 2, iv.length / 2);
+            cipher = getDecryptionCipher(iv2, refreshToken, salt);
+            final String supersedingRefreshToken = new String(cipher.doFinal(encryptedSupersedingRefreshToken), StandardCharsets.UTF_8);
+            listener.onResponse(new Tuple<>(prependVersionAndEncodeAccessToken(refreshTokenStatus.getVersion(), supersedingAccessToken),
+                prependVersionAndEncodeRefreshToken(refreshTokenStatus.getVersion(), supersedingRefreshToken)));
+        } catch (GeneralSecurityException | IOException e) {
+            logger.warn("Could not get stored superseding token values", e);
+            listener.onFailure(invalidGrantException("could not refresh the requested token"));
+        }
     }
 
-    private void getSupersedingTokenDocAsync(RefreshTokenStatus refreshTokenStatus, ActionListener<GetResponse> listener) {
-        final String supersedingDocReference = refreshTokenStatus.getSupersededBy();
-        if (supersedingDocReference.startsWith(securityTokensIndex.aliasName() + "|")) {
-            // superseding token doc is stored on the new tokens index, irrespective of where the superseded token doc resides
-            final String supersedingDocId = supersedingDocReference.substring(securityTokensIndex.aliasName().length() + 1);
-            getTokenDocAsync(supersedingDocId, securityTokensIndex, listener);
-        } else {
-            assert false == supersedingDocReference
-                    .contains("|") : "The superseding doc reference appears to contain an alias name but should not";
-            getTokenDocAsync(supersedingDocReference, securityMainIndex, listener);
-        }
+    /*
+     * Encrypts the values of the superseding access token and the refresh token, using a key derived from the superseded refresh token.
+     */
+    Tuple<String, String> encryptSupersedingTokens(String supersedingAccessToken, String supersedingRefreshToken,
```

Comment:
Discarded the idea of using one time pads for encryption as originally discussed because of requirements around FIPS 140-2 compliance

Reply:
Good call!